src/tests/dbus-tests/test_btrfs.py

```diff
@@ -0,0 +1,293 @@
+            name = out.strip().split(' ')[-1]
+            self.assertEqual(name, 'test_sub1')
+
+            # list all subvolumes
+            subs, num = dev.obj.GetSubvolumes(False, self.no_options,
+                                              dbus_interface=self.iface_prefix + '.Filesystem.BTRFS')
+            self.assertEqual(num, 1)
+            self.assertEqual(subs[0][2], 'test_sub1')  # name (path) is the third item
+
+            # delete the subvolume
+            dev.obj.RemoveSubvolume('test_sub1', self.no_options,
+                                    dbus_interface=self.iface_prefix + '.Filesystem.BTRFS')
+            _ret, out = self.run_command('btrfs subvolume list %s' % mnt)
+            self.assertFalse(out)  # no subvolume -> empty output
+
+            # list all subvolumes
+            _subs, num = dev.obj.GetSubvolumes(False, self.no_options,
+                                               dbus_interface=self.iface_prefix + '.Filesystem.BTRFS')
+            self.assertEqual(num, 0)
+
+    def test_add_remove_device(self):
+        dev1, dev2 = self._get_devices(2)
+        self.addCleanup(self._clean_format, dev1.obj)
+        self.addCleanup(self._clean_format, dev2.obj)
+
+        manager = self.get_object('/Manager')
+        manager.CreateVolume([dev1.path],
+                             'test_add_remove', 'single', 'single',
+                             self.no_options,
+                             dbus_interface=self.iface_prefix + '.Manager.BTRFS')
+        fstype = self.get_property(dev1.obj, '.Block', 'IdType')
+        fstype.assertEqual('btrfs')
+
+        # add second device to the volume
+        with self._temp_mount(dev1.path):
+            dev1.obj.AddDevice(dev2.path, self.no_options,
+                               dbus_interface=self.iface_prefix + '.Filesystem.BTRFS')
+            fstype = self.get_property(dev2.obj, '.Block', 'IdType')
+            fstype.assertEqual('btrfs')
+
+        # check filesystem type of the new device
+        fstype = self.get_property(dev2.obj, '.Block', 'IdType')
+        fstype.assertEqual('btrfs')
+
+        # check number of devices
+        dbus_devs = self.get_property(dev1.obj, '.Filesystem.BTRFS', 'num_devices')
+        dbus_devs.assertEqual(2)
+
+        _ret, out = self.run_command('btrfs filesystem show %s' % dev1.path)
+        self.assertIn('Total devices 2 FS', out)
+        self.assertIn('path %s' % dev1.path, out)
+        self.assertIn('path %s' % dev2.path, out)
+
+        # remove the second device
+        with self._temp_mount(dev1.path):
+            dev1.obj.RemoveDevice(dev2.path, self.no_options,
+                                  dbus_interface=self.iface_prefix + '.Filesystem.BTRFS')
+            fstype = self.get_property(dev2.obj, '.Block', 'IdType')
+            fstype.assertFalse()
+
+        # check number of devices
+        # storaged currently reports wrong number of devices (2)
```

Comment:
Please mark this comment with ``XXX:`` or ``BUG:`` so that we can easily find it in the future.